### PR TITLE
[TensorRT/Filter] Add a sub-plugin filter for TensorRT @open sesame 10/5 11:04

### PR DESCRIPTION
**Pull request by again4you**
This PR newly adds an sub-plugin filter for TensorRT and its related test cases.
It is still the WIP phase.

Signed-off-by: Sangjung Woo <sangjung.woo@samsung.com>

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed []Skipped
2. Run test: [*]Passed [ ]Failed []Skipped

### v2
* Add cumtom parameter 'input_rank', which is required by UFF model.
  * [nvinfer1::Dims Class Reference](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_dims.html)
  * [Code](https://github.com/nnstreamer/nnstreamer/pull/2771/commits/55b8ee60638ecd54f3915dfe731c82e327bdc4bc#diff-5329ab45a3e15b7fd4b642443b24c10fR288)
  * Current approach
   ```
   ...
   tensor_filter framework=tensorrt model=${PATH_TO_MODEL} \
   input=28:28:1 inputtype=float32 custom=input_rank:3 inputname=in output=10 outputtype=float32 outputname=out
   ...
   ```
* Use the unified memory if it is supported.
* Add MNIST test case and check its working

### v1
* First draft
* Support UFF (universal framework format) model, which is converted from TensorFlow.

### CODE REVIEW

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorrt.cc [hunk 1]
```diff
@@ -0,0 +1,643 @@
...
+using Severity = nvinfer1::ILogger::Severity;
+
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+
+/** @brief a global object of ILogger */
+class Logger : public nvinfer1::ILogger
+{
+	void log(Severity severity, const char* msg) override
+	{
+    switch (severity) {
```
Comment (dongju-chae) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorrt.cc:45]: Same indent issue :)

**Review by dongju-chae — APPROVED** [review]
LGTM

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorrt.cc [hunk 2]
```diff
@@ -0,0 +1,643 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+/**
+ * GStreamer Tensor_Filter, TensorRT Module
+ * Copyright (C) 2020 Dongju Chae <dongju.chae@samsung.com>
+ * Copyright (C) 2020 Sangjung Woo <sangjung.woo@samsung.com>
+ */
+/**
+ * @file   tensor_filter_tensorrt.cc
+ * @date   24 Sep 2020
+ * @brief  Mediapipe module for tensor_filter gstreamer plugin
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorrt.cc:10]: This has nothing to do with Mediapipe.

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorrt.cc [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+ * @see    http://github.com/nnstreamer/nnstreamer
+ * @author Dongju Chae <dongju.chae@samsung.com>
+ *         Sangjung Woo <sangjung.woo@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ * This is the per-NN-framework plugin (TensorRT) for tensor_filter.
+ *
+ * @note Only support UFF (universal framework format) file as inference model.
+ * @note When building the pipeline, cumtom parameter 'input_rank' is required, which is matched to UFF model file.
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorrt.cc:19]: cumtom parameter --> custom property
Add a usage example for custom properties here, too.

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorrt.cc [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+  template <typename T>
+  void operator()(T* obj) const
+  {
+    if (obj)
+      obj->destroy();
+  }
+};
+
+/**
+ * @brief class definition for TensorRT custom filter
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorrt.cc:74]: This is not a custom filter.

**Comment by myungjoo** [comment]
With #2878 , you may implement this as a subclass of C++ tensor-filter, which is expected to reduce the line count.

**Comment by again4you** [comment]
> With #2787 , you may implement this as a subclass of C++ tensor-filter, which is expected to reduce the line count.

OK. After adding `rank` information to `TensorInfo`, I will update this PR.

**Comment by myungjoo** [comment]
> > With #2787 , you may implement this as a subclass of C++ tensor-filter, which is expected to reduce the line count.
> 
> OK. After adding `rank` information to `TensorInfo`, I will update this PR.

As we've discussed in Thursday, please make "inputrank" property in tensor-filter common for all subplugins who want it, Rank cannot exist in TensorInfo (it cannot exist in GstCap. Rank will break GstCap negotiation process).

**Pull request closed without merging**